dev-notes/minimal-os-for-fobnail.md

```diff
@@ -0,0 +1,569 @@
+| USB host driver         | Required for communication with Fobnail                                        |
+| USB EEM driver          | Required for communication with Fobnail                                        |
+| Network Stack           | Required for communication with Fobnail                                        |
+| TPM driver              | Required to perform attestation                                                |
+| Bootloader Capabilities | Required to boot target OS                                                     |
+| C library               | [fobnail-attester](https://github.com/fobnail/fobnail-attester) is writen in C |
+| Bootable by SKL         | Whether OS can be loaded by TrenchBoot SKL without SKL or OS modification      |
+
+These are another features which are taken into account (soft requirements).
+
+| Feature                  | Description                                                               |
+| ------------------------ | ------------------------------------------------------------------------- |
+| Microkernel              | Microkernels are more secure and are preferred                            |
+| OS portability           | Required to avoid rebuilding minimal OS for each device                   |
+| CPU Architecture support | OS supported architectures, mostly we consider x86, ARM, RISC-V and POWER |
+
+### OS score counting
+
+The table at the bottom of this document contains summary of features of all
+OSes together with theirs scores. Each score is computed using the following
+rules
+
+- If feature is present then +1
+- If feature is not present then 0, or if feature would be hard to implement
+  then -1
+- If this is a hard requirement then multiple result by 2
+
+## Different OSs propositions
+
+The research effect is presented below. 4 systems were considered:
+
+* Zephyr
+* Xous
+* seL4
+* Linux
+
+Each of them has a short description, an analysis of the launch in DLME, and the
+possibilities and potential problems that will have to be addressed for the
+Fobnail Token to be functional.
+
+### Zephyr
+
+Zephyr is a well-known and supported embedded RTOS with a monolithic kernel
+written in C. It supports various devices ranging from Cortex-M to x86 CPUs.
+It's main benefits are IPv4, CoAP, and USB (including USB EEM) support which are
+required for communication with Fobnail token. Other benefits are listed
+[here](https://www.zephyrproject.org/benefits/).
+
+#### Running in DLME
+
+Secure launch uses a component called
+[SKL](https://github.com/TrenchBoot/secure-kernel-loader) (Secure Kernel Loader)
+which is responsible for booting OS in DLME (Zephyr in that case). Zephyr boots
+using Multiboot1 but SKL does not support it (only Multiboot2 is supported). For
+SKL to be able to boot Zephyr, Zephyr must gain Multiboot2 support.
+
+Multiboot1 is hard measure due to its tags being scattered around in memory. To
```

Comment:
```suggestion
Multiboot1 is hard to measure due to its tags being scattered around in memory. To
```

Reply:
Fixed

Reply:
I will resolve this thread as the change is small and has been applied.